File: 백준/Gold/1766. 문제집/문제집.py
#Authored by : chjh2129
import sys
from heapq import heappush as hpush, heappop as hpop

'''
  우선순위 큐로 위상 정렬을 수행합니다.

  낮은 번호의 문제 즉, 쉬운 문제를 먼저 풀도록 합니다.
'''

input = sys.stdin.readline

def main() -> None:
  # input & init
  n, m = map(int, input().split())
  
  adj = [ [] for _ in range(n+1) ]
  deg = [ 0 for _ in range(n+1) ]

  for u, v in (map(int, input().split()) for _ in range(m)):
    adj[u].append(v)
    deg[v] += 1
  
  # topological sort
  pq: list[int] = []
  ans: list[int] = []

  for i in range(1, n+1):
    if not deg[i]:
      hpush(pq, i)

  while pq:
    cur = hpop(pq)
    ans.append(cur)

    for nxt in adj[cur]:
      deg[nxt] -= 1
      if not deg[nxt]:
        hpush(pq, nxt)
  
  # output
  print(*ans)

sys.exit(main())

File: 백준/Gold/1766. 문제집/문제집.java
// Authored by : chjh212
import java.io.*;
import java.util.*;

public class Main {
  static int N, M;
  static PriorityQueue<Integer> pq = new PriorityQueue<>();
  static List<List<Integer>> adj = new ArrayList<>();
  static int[] deg;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input & init
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());
    
    for(int i = 0; i <= N; i++) adj.add(new ArrayList<>());
    deg = new int[N+1];
    
    while(M-- > 0){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());

      adj.get(u).add(v);
      deg[v]++;
    }

    // topological sort
    for(int i = 1; i <= N; i++){
      if(deg[i] == 0){
        pq.add(i);
      }
    }

    while(!pq.isEmpty()){
      var cur = pq.poll();
      
      sb.append(cur+" ");
      
      for(var nxt : adj.get(cur)){
        deg[nxt]--;
        if(deg[nxt] == 0) pq.add(nxt);
      }
    }

    // output
    System.out.println(sb.toString());
  }
}